### DISCUSSION TITLE
API First, TDD and Documentation first approach

### DISCUSSION THREAD

**Discussion by wtrocki** [aerogear/charmil#61] [General]
With fast paced upstream work developers often forget that we sometimes write software for the people not machines. 
Charmil is the example of such project as as framework needs to have good UX and clear integration paths.
Without that it will not succeed as upstream project.

To prevent from doing that we need to have developer centric focus when we coding anything - we do not write code for mentors, we do not write code to satisfy our requirements - we write them for end users to see value. 

# How?

Some propositions for the flows that worked for me in the past.

1. API first

When working on PR define end user API first with proper documentation first and create draft PR. 
You can provide stub that returns some result if needed. In case we wrapping something we might also 
reference what we going to wrap and what arguments will be or not be exposed.

2. Documentation First

Try to write documentation before you start coding. Example code is also documentation.
In practice the smaller example (something that doesn't have much unrelated code) then better for the users.

3. (Optional) Try to write tests to validate outputs.

Unt tests often force us to validate if we handle all outputs properly.
Use your judgment when creating tests - do not write tests for wrapper but if you doing some logic or have helper having nice tests is good. Avoid extensive mocking or large codebase in tests (everything that will slow down development later)

**Comment by jackdelahunt** [comment]
Can you give more detail on this documentation stage as I am not sure what you mean by documentation here. The reason for my confusion is that I would think that making changes to external documentation(Github wiki/pages) before the pr that those changes depend on is merged wouldn't happen. And if that is the case I am interested in what you mean by it here.

Unless if this means changing the docs in the same pr, then making changes to the code base in subsequent commits for that pr?

**Reply by wtrocki** [reply]
> I would think writing to public documentation before a pr is merged adding that feature wouldn't happen. 

Can you edit this sentence to make it more clear what is the question?

To expand:
I meant that PR contains also documentation, examples, go docs (comments in code) and everything that makes it fully usable and releasable. Using master/main for user documentation is antipattern for repositories that have active development and long release lifecycle. You should have website or github pages for that and shelve this from gh-pages anyway. See #62 

A person who writes code is the best person to document it and the best time to document it is when you are fresh with code.
Imagine being intern that was told to document or create unit tests for someones else code and having "amazing" experience when doing it :). This is what we trying to avoid.

With young developers I see that they often code faster than they think so here I'm proposing to start with documentation first - this will force you think about specific problem from user perspective rather from - that will be cool code to write or that will be cool thing to add.

**Reply by jackdelahunt** [reply]
> I meant that PR contains also documentation, examples, go docs (comments in code) and everything that makes it fully usable and releasable. 

Okay from this I understand, what we were thinking as documentation was different as I was just thinking about docs on a website of github pages. But saying we need to document first in reference to go docs and examples I agree with.

**Comment by jjkiely** [comment]
I like the idea of going with the documentation first approach. Especially with the project at the early stages it is in now, there is no better time to start. I think it's a good idea to discuss as a team what the intended feature set of the project is and how it will be usable; i.e What do we want charmil to do? How would an interested party make use of charmil?

I think it would be beneficial to discuss these questions and others, I think a clear plan and clear docs can make for easier development as there is something concrete to have as a goal. Which would lead to a clearer understanding of what you are tackling when working on a particular issue. Also if docs are being written concurrently with code, after the initial creation of a doc for a given feature the amount of effort to maintain these docs is considerably lower for subsequent changes/updates.

Having good docs will also make the project far more useful from the end user's point of view. It's infinitely harder to build a community without good documentation. If someone was to come along and read our current README they would say "hey that's a pretty good idea" whereas if we had more detailed documentation (Planned future features, intended use cases etc) and perhaps a simple coded and documented example of how to create plugins for your cli, I feel like the person would be far more likely to see just how useful charmil could be, and may even encourage contribution on their part.

**Reply by jjkiely** [reply]
Another point just crossed my mind also. Consistent documentation also helps to prevent knowledge silos (where only the person who implemented a feature knows how it works), if code written by someone else is readable and well documented it is far easier to extend/improve on it . It can be very difficult to know where to start when trying to add to existing code without good documentation (both comments in code and outside of it)

**Reply by wtrocki** [reply]
> What do we want charmil to do? How would an interested party make use of charmil?

That really makes upstream project upstream - usability and general audience. No one will bait an eye on some side project you have been doing in the open that resolving some problem only you know about :)

> Having good docs

In practice I have seen many people pushing back on this because they did not want to put documentation without knowing that their approach is right, but after forcing them to write docs I instantly stopped seeing bad approaches - they were just rejected at the time when docs is created. Writing docs forces people to build simple and usable solutions (as you do not want to spend ages documenting stuff and need to be being very clear about the problem you solve). This keeps some developers in line to not add extra useful things (often buggy and undocumented)

**Reply by jjkiely** [reply]
> Writing docs forces people to build simple and usable solutions (as you do not want to spend ages documenting stuff and need to be being very clear about the problem you solve)

100% agree, thinking from the end user's perspective while writing docs is extremely important to develop something that will actually be used. If it's too intimidating people will look elsewhere

**Comment by maksymvavilov** [comment]
Reflecting on my personal experience with projects I had: 
Documentation first approach was good only for projects where you know exactly what and how from the first moment of it. Reminds of the `Waterfall Model` and things. 
However, while developing in changing environment or one of the parts of the development process is to research how certain things must be developed, creating documentation at the beginning is just silly work. 
I would agree, that docs should be present, but when choosing what shall come first - API or docs - I would vote for the API. Otherwise, it would be a situation when "before starting coding and producing results I must spend a week figuring out exactly how the code will work". 
While with an API first approach, by defining an API you would say "this is what the code will do" and explain "how" part after you finished the coding. 
If I'm not missing anything for the `fast-paced upstream work` more agile approach is a better match :thinking:

**Reply by wtrocki** [reply]
I wish I had mind sync with people. API first, Docs first (as for documenting what you have done in API) first and full scale docs second was what I have intended to pass. I guess that numbering above doesn't reflect that clearly. Will change it now

**Reply by wtrocki** [reply]
> before starting coding and producing results I must spend a week figuring out exactly how the code will work".

Oh.. you probably want to take part of this discussion then:
https://github.com/aerogear/charmil/discussions/60

:)

**Reply by maksymvavilov** [reply]
> > before starting coding and producing results I must spend a week figuring out exactly how the code will work".
> 
> Oh.. you probably want to take part of this discussion then:
> #60
> 
> :)

Well, the "sually worked for me" seems pretty much like what I have in mind. Define boundaries of the project (it will take you chat messages and turn every "O" into :open_mouth: ), then implement it, Document the code so other people might also use it, and publish :thinking:

**Reply by wtrocki** [reply]
Good stuff!

**Comment by ankithans** [comment]
@ all, thanks for the #60 & #61 discussions. This will really allow me to follow the good practices. You will definitely see changes in my upcoming contributions 🤞

docs first is relatable, When I don't know what to do, I try to write some trash code before implementing the feature, to check whether it would work. I now tend to write whatever comes to me in text, which aids me in swiftly dismiss or accept ideas. And this can be later refined to be called as docs!

**Comment by wtrocki** [comment]
Reminder about this topic @ankithans @namit-chandwani . I will stop reviewing PRs that without instructions and basic information on what problem we resolve and how to use your code.

**Comment by wtrocki** [comment]
Extract from today's discussion:

- Without discussion on user facing API and values we want to provide  to devs we cannot really start working on PR.
- It is faster and easier to talk about code than write it.
- Writing code without having in mind picture how it will be used and executed is very risky and can produce random results.
- Exposing the mentor to the code that we did not discuss prolongs the process and causes numerous of comments.
It affects students as they need to deal with number of comments on PR, but overall PR quality might not improve 
- When starting project we cannot deal with all problems at the same time - creating follow up issues is desired. 
Sometimes we might create PR with just top level API and some minimal docs and merge it. Then work on implementation separately. 
Doing opposite - have large chunks of code without user facing API will most like cause problems.

**Reply by ankithans** [reply]
> Sometimes we might create PR with just top level API and some minimal docs and merge it. Then work on implementation separately.

This looks good and effective.
> Without discussion on user facing API and values we want to provide to devs we cannot really start working on PR.

Yes I have experienced this in previous PR. I will gather the requirements first, then start to code.

**Reply by wtrocki** [reply]
To not be very strict.. Sometimes you need to try things out before you define API. That is normal, but I feel like we should have conversation/think about API first and also investigate more - linters are pretty common area and there are tons of blog posts explaining how most of them work.

**Reply by ankithans** [reply]
yep, thanks! Let me do some research over this, I will discuss & follow up
